File: api/app/infrastructure/hardware/daemon_client.py
"""Client for the C hardware daemon (``mjqbe-daemon``).

Same wire protocol as the daemon and the Rust client: one JSON object per line
over a Unix socket. One short-lived connection per request.
"""
import json
import os
import socket

_TIMEOUT = 3.0


class HardwareUnavailable(RuntimeError):
    """The daemon socket could not be reached."""


class DaemonError(RuntimeError):
    """The daemon replied with ``ok: false``."""


def _socket_path() -> str:
    return os.getenv("DAEMON_SOCKET", "/run/mjqbe/daemon.sock")


def _request(cmd: str, **params) -> dict:
    payload = {"cmd": cmd, **params}
    try:
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.settimeout(_TIMEOUT)
        s.connect(_socket_path())
    except OSError as exc:
        raise HardwareUnavailable(str(exc)) from exc

    try:
        s.sendall((json.dumps(payload) + "\n").encode())
        buf = b""
        while b"\n" not in buf:
            chunk = s.recv(4096)
            if not chunk:
                break
            buf += chunk
    finally:
        s.close()

    try:
        resp = json.loads(buf.decode().splitlines()[0])
    except (ValueError, IndexError) as exc:
        raise DaemonError(f"bad daemon response: {buf!r}") from exc

    if not resp.get("ok"):
        raise DaemonError(resp.get("error", "daemon error"))
    return resp.get("data", {})


def info() -> dict:
    return _request("info")


def gpio_set(pin: int, value: int) -> dict:
    return _request("gpio_set", pin=int(pin), value=1 if value else 0)


def gpio_get(pin: int) -> dict:
    return _request("gpio_get", pin=int(pin))


def relay_set(relay: int, state: int) -> dict:
    return _request("relay_set", relay=int(relay), state=1 if state else 0)


def led_set(r: int, g: int, b: int) -> dict:
    return _request("led_set", r=int(r), g=int(g), b=int(b))


_AV_ACTIONS = {"tv_on", "tv_off", "tv_toggle", "ps4_on", "ps4_off"}


def av_status() -> dict:
    return _request("av_status")


def av_cec(action: str) -> dict:
    if action not in _AV_ACTIONS:
        raise DaemonError(f"invalid AV action: {action}")
    return _request("cec_send", action=action)

File: native/core/src/infrastructure/hardware/daemon_client.rs
//! Client for the C hardware daemon (`mjqbe-daemon`).
//!
//! Same wire protocol as the daemon: one JSON object per line over a Unix
//! socket. One short-lived connection per request — the daemon is local and
//! requests are rare (GPIO toggles), so this keeps the client trivial and
//! stateless.

use std::time::Duration;

use serde_json::{json, Value};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixStream;

use crate::domain::CoreError;

pub struct DaemonClient {
    socket_path: String,
}

impl DaemonClient {
    pub fn new(socket_path: impl Into<String>) -> Self {
        Self {
            socket_path: socket_path.into(),
        }
    }

    /// Reads `DAEMON_SOCKET`, else `/run/mjqbe/daemon.sock`.
    pub fn from_env() -> Self {
        Self::new(
            std::env::var("DAEMON_SOCKET").unwrap_or_else(|_| "/run/mjqbe/daemon.sock".to_string()),
        )
    }

    async fn request(&self, cmd: &str, mut params: Value) -> Result<Value, CoreError> {
        params["cmd"] = json!(cmd);

        let connect = UnixStream::connect(&self.socket_path);
        let stream = tokio::time::timeout(Duration::from_secs(8), connect)
            .await
            .map_err(|_| CoreError::HardwareUnavailable)?
            .map_err(|_| CoreError::HardwareUnavailable)?;

        let (rd, mut wr) = stream.into_split();
        let mut line = serde_json::to_vec(&params).expect("serialize daemon request");
        line.push(b'\n');
        wr.write_all(&line)
            .await
            .map_err(|e| CoreError::Internal(e.to_string()))?;
        wr.flush().await.ok();

        let mut reader = BufReader::new(rd);
        let mut resp = String::new();
        tokio::time::timeout(Duration::from_secs(8), reader.read_line(&mut resp))
            .await
            .map_err(|_| CoreError::Internal("daemon timeout".into()))?
            .map_err(|e| CoreError::Internal(e.to_string()))?;

        let v: Value = serde_json::from_str(resp.trim())
            .map_err(|e| CoreError::Internal(format!("bad daemon response: {e}")))?;

        if v.get("ok").and_then(Value::as_bool).unwrap_or(false) {
            Ok(v.get("data").cloned().unwrap_or(Value::Null))
        } else {
            let msg = v
                .get("error")
                .and_then(Value::as_str)
                .unwrap_or("daemon error")
                .to_string();
            Err(CoreError::Internal(msg))
        }
    }

    pub async fn info(&self) -> Result<Value, CoreError> {
        self.request("info", json!({})).await
    }

    pub async fn gpio_set(&self, pin: u8, value: bool) -> Result<Value, CoreError> {
        self.request("gpio_set", json!({ "pin": pin, "value": value as u8 }))
            .await
    }

    pub async fn gpio_get(&self, pin: u8) -> Result<bool, CoreError> {
        let d = self.request("gpio_get", json!({ "pin": pin })).await?;
        Ok(d.get("value").and_then(Value::as_i64).unwrap_or(0) != 0)
    }

    pub async fn relay_set(&self, relay: u8, state: bool) -> Result<Value, CoreError> {
        self.request("relay_set", json!({ "relay": relay, "state": state as u8 }))
            .await
    }

    pub async fn led_set(&self, r: u8, g: u8, b: u8) -> Result<Value, CoreError> {
        self.request("led_set", json!({ "r": r, "g": g, "b": b }))
            .await
    }

    pub async fn cec_send(&self, action: &str) -> Result<Value, CoreError> {
        self.request("cec_send", json!({ "action": action })).await
    }

    pub async fn av_status(&self) -> Result<Value, CoreError> {
        self.request("av_status", json!({})).await
    }
}
